dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/util/CpuInfo.java

```diff
@@ -345,4 +370,59 @@ public double getSystemCpuLoad() {
             }
         }
     }
+
+    private static int currentNumberOfProcessors = -1;
```

Comment:
What is the point of this field?  Why can't it just be a local variable below that calls AVAILABLE_PROCESSORS.get()?  I don't know why the taskSync is needed as well.  If it is only there to protect currentNumberOfProcessors and you make that a local field to the method, then I think it isn't needed.

Reply:
The sync block and currentNumberOfProcessors allows us to avoid calling back the listeners when the timer wakes up but the number or processors hasn't changed. Also when we restore we can have multiple threads potentially calling out to notifyListeners. This is because even though we cancel the previous task, the behavior shows it will "catch up" and fire however many times it missed while the process was gone. This is how I solve those issues.

Reply:
My point is having a instance scoped value doesn't make sense to me since you can easily just get it from the AVAILABLE_PROCESSORS field

Reply:
I am thinking code something like:

```
        public void run() {
            // find available processors
            int runtimeAvailableProcessors = Runtime.getRuntime().availableProcessors();
            int fileSystemAvailableProcessors = getAvailableProcessorsFromFilesystem();

            int newAvailableProcessors;
            if (fileSystemAvailableProcessors <= 0 || fileSystemAvailableProcessors > runtimeAvailableProcessors) {
                newAvailableProcessors = runtimeAvailableProcessors;
            } else  {
                newAvailableProcessors = runtimeAvailableProcessors;
            }

            int currentNumberOfProcessors = AVAILABLE_PROCESSORS.get();
            if (currentNumberOfProcessors != newAvailableProcessors) {
                // if another thread runs the task and set it to a different value before
                // we do, then ignore this update since they should have set it to the same
                // value we were going to set it to likely.
                if (AVAILABLE_PROCESSORS.compareAndSet(currentNumberOfProcessors, newAvailableProcessors)) {
                    notifyListeners(newAvailableProcessors);
                }
            }
        }
```

Reply:
Agreed on getting rid of currentNumberOfProcessors. I still think the sync block is needed so we don't get false notifications when the number hasn't actually changed.

Reply:
This is why I used the compareAndSet above.  I have additional concerns we are doing a lot of work with the getting of the current number of processors.  I am thinking we should possibly have some logic to check the last time we ran to make sure we are not doing the catchup of running 100s of times and calculating the number or processors which may get expensive.  It would be better to check since last time we ran and only run if it hasn't be at least X number of minutes or seconds or something.